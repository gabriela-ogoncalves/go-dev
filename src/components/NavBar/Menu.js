import React, { Fragment, useState, useContext, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import swal from 'sweetalert';

import Box from '@mui/material/Box';
import Avatar from '@mui/material/Avatar';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import ListItemIcon from '@mui/material/ListItemIcon';
import Divider from '@mui/material/Divider';
import IconButton from '@mui/material/IconButton';
import Tooltip from '@mui/material/Tooltip';
import Logout from '@mui/icons-material/Logout';

import Context from '../../Context';
import AuthService from '../../services/auth.js';
import MenuIcon from './MenuIcon';

export default function AccountMenu() {
  const [anchorEl, setAnchorEl] = useState(null);
  const [isLogged, setIsLogged] = useContext(Context);
  const open = Boolean(anchorEl);
  const navigate = useNavigate();

  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  useEffect(() => {},[isLogged]); 
    
  return (
    <>
      <Box sx={{ display: 'flex', alignItems: 'center', textAlign: 'center', marginRight: '1rem' }}>
        {
          isLogged 
          ? (
            <Tooltip title="Minha conta">
              <IconButton
                onClick={handleClick}
                size="small"
                sx={{ ml: 2 }}
                aria-controls={open ? 'account-menu' : undefined}
                aria-haspopup="true"
                aria-expanded={open ? 'true' : undefined}
              >
                <Avatar sx={{ width: 45, height: 45, bgcolor:'var(--go-dev-border)' }} />
              </IconButton>
            </Tooltip>
          ) : (
            <Tooltip title="Minha conta">
              <IconButton
                onClick={handleClick}
                size="small"
                sx={{ ml: 2 }}
                aria-controls={open ? 'account-menu' : undefined}
                aria-haspopup="true"
                aria-expanded={open ? 'true' : undefined}
              >
                <Avatar
                  sx={{ width: 45, height: 45, bgcolor:'var(--go-dev-border)' }} 
                >
                  <MenuIcon />
                </Avatar>
              </IconButton>
          </Tooltip>
          )
        }
      </Box>
      {
        isLogged &&
        <Menu
          anchorEl={anchorEl}
          id="account-menu"
          open={open}
          onClose={handleClose}
          onClick={handleClose}
          PaperProps={{
            elevation: 0,
            sx: {
              overflow: 'visible',
              filter: 'drop-shadow(0px 2px 8px rgba(0,0,0,0.32))',
              mt: 1.5,
              '& .MuiAvatar-root': {
                width: 32,
                height: 32,
                ml: -0.5,
                mr: 1,
              },
              '& .MuiList-root': {
                fontFamily: 'var(--open-sans-font)'
              },

              '&:before': {
                content: '""',
                display: 'block',
                position: 'absolute',
                top: 0,
                right: 14,
                width: 10,
                height: 10,
                bgcolor: 'background.paper',
                transform: 'translateY(-50%) rotate(45deg)',
                zIndex: 0,
              },
            },
          }}
          transformOrigin={{ horizontal: 'right', vertical: 'top' }}
          anchorOrigin={{ horizontal: 'right', vertical: 'bottom' }}
        >
          <MenuItem disabled>
            Bem-vindo(a), <strong>{isLogged.username}</strong>!
          </MenuItem>
          <Divider />

          <MenuItem onClick={() => {
            handleClose();
            navigate('/profile', {replace: true});
          }}>
            Meu Perfil
          </MenuItem>
          <MenuItem onClick={() => {
            handleClose();
            navigate('/trilhas', {replace: true});
          }}>
            Trilhas
          </MenuItem>
          <MenuItem onClick={() => {
            handleClose();
            navigate('/certificados', {replace: true});
          }}>
            Certificados
          </MenuItem>
          <Divider />
          <MenuItem onClick={() => {
            swal({
              title: 'Tem certeza que deseja sair?',
              text: 'Temos muito mais trilhas que ainda podem te interessar! Que tal dar uma olhadinha?',
              icon: 'warning',
              buttons: ['Cancelar', 'Sair'],
            })
            .then((confirm) => {
              if (confirm) {
                AuthService.logout();
                setIsLogged(false);
                handleClose();
                navigate('/', {replace: true});
              }
            });
          }}
          >
            <ListItemIcon>
              <Logout fontSize="small" />
            </ListItemIcon>
            Sair
          </MenuItem>
        </Menu>
      }
      {
        !isLogged &&
        <Menu
          anchorEl={anchorEl}
          id="account-menu"
          open={open}
          onClose={handleClose}
          onClick={handleClose}
          PaperProps={{
            elevation: 0,
            sx: {
              overflow: 'visible',
              filter: 'drop-shadow(0px 2px 8px rgba(0,0,0,0.32))',
              mt: 1.5,
              '& .MuiAvatar-root': {
                width: 32,
                height: 32,
                ml: -0.5,
                mr: 1,
              },
              '& .MuiList-root': {
                width: 160
              },
              '&:before': {
                content: '""',
                display: 'block',
                position: 'absolute',
                top: 0,
                right: 14,
                width: 10,
                height: 10,
                bgcolor: 'background.paper',
                transform: 'translateY(-50%) rotate(45deg)',
                zIndex: 0,
              },
            },
          }}
          transformOrigin={{ horizontal: 'right', vertical: 'top' }}
          anchorOrigin={{ horizontal: 'right', vertical: 'bottom' }}
        >

          <MenuItem onClick={() => {
            handleClose();
            navigate('/trilhas', {replace: true});
          }}>
            Trilhas
          </MenuItem>
          <MenuItem onClick={() => {
            handleClose();
            navigate('/cadastro', {replace: true});
          }}>
            Cadastrar
          </MenuItem>
          <MenuItem onClick={() => {
            handleClose();
            navigate('/login', {replace: true});
          }}>
            Login
          </MenuItem>
        </Menu>
      }
    </>
  );
}