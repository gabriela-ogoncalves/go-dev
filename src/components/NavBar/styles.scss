.navigation {
    position: relative;
    display: flex;
    align-items: center;
    height: 90px;
    width: 100%;
    padding: 0.5rem 0rem;
    box-shadow: 0 2px 2px 2px var(--go-dev-border);
    background-color: var(--go-dev-background);
    font-family: var(--bungee-font);
  }

  .brand-name {
    text-decoration: none;
    color: black;
    font-size: 1.3rem;
    margin-left: 1rem;
  }

  .logo-image {
    width: 149px;
    height: 72px;
  }
  
  .navigation-menu {
    margin-left: auto;

    ul {
        display: flex;
        padding: 0;
    }

    li {
        list-style-type: none;
        margin: 0 0.5rem;
        border-radius: 10px;

        &.active {
            background-color: #734a9975;
          }
      
          &:hover {
            background-color: #734a9957;
          }
          
        a {
            display: flex;
            align-items: center;
            justify-content: center;
            color: inherit;
            text-decoration: none;
            height: 100%;
            width: 100%;
            padding: 1rem;
            color: var(--go-dev-text);        
        }
      }
  }
  
  .hamburger {
    border: 0;
    height: 40px;
    width: 40px;
    padding: 0.5rem;
    border-radius: 50%;
    background-color: var(--go-dev-border);
    cursor: pointer;
    transition: background-color 0.2s ease-in-out;
    position: absolute;
    top: 50%;
    right: 25px;
    transform: translateY(-50%);
    display: none;

    &:hover {
        opacity: 0.6;
    }
  }
  
  @media screen and (max-width: 700px) {
    .hamburger {
      display: block;
    }
  
    .navigation-menu {
        ul {
            flex-direction: column;
            position: absolute;
            top: 75px;
            left: 0;
            width: 100%;
            background-color: white;
            border-top: 1px solid black;
            display: none;
        }

        li {
            text-align: center;
            margin: 0;

            &:hover {
                background-color: #eee;
            }

            a {
                color: var(--go-dev-text);
                width: 100%;
                padding: 1.5rem 0;
            }
        }
    }
  
    .navigation-menu.expanded ul {
      display: block;
      z-index: 1;
    }
  }
  