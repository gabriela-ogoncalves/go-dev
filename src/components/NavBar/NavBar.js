import { useState } from 'react';
import { Link, useMatch, useResolvedPath } from 'react-router-dom';

import logo from '../../assets/logo/go-dev-logo.png';
import MenuIcon from './MenuIcon';

import './styles.scss';

const NavBar = () => {
  const [isNavExpanded, setIsNavExpanded] = useState(false);

  const toggleNav = () => {
    setIsNavExpanded(!isNavExpanded);
  };

  return (
    <article data-testid='nav-bar'>
      <nav className='navigation'>
          <Link
              to='/'
              data-testid='go-dev-logo'
              className='brand-name'
          >
              <img src={logo} alt='Logo Go Dev' className="logo-image" />
          </Link>

        <button
          className='hamburger'
          onClick={toggleNav}
          type='button'
          name='menu'
        >
          <MenuIcon />
        </button>
        <div
          className={
            isNavExpanded ? 'navigation-menu expanded' : 'navigation-menu'
          }
        >
          <ul>
            <CustomLink onClick={toggleNav} role='link' href='/trilhas'>Trilhas</CustomLink>
            <CustomLink onClick={toggleNav} role='link' href='/certificados'>Certificados</CustomLink>
            <CustomLink onClick={toggleNav} role='link' href='/cadastro'>Cadastro</CustomLink>
            <CustomLink onClick={toggleNav} role='link' href='/login'>Login</CustomLink>          
          </ul>
        </div>
      </nav>
    </article>
  );
};

const CustomLink = ({ href, children, ...props }) => {
    const path = useResolvedPath(href);
    const isActive = useMatch({ path: path.pathname, end: true });

    return (
        <li className={isActive ? 'active' : ''}>
        <Link to={href} {...props}>
            {children}
        </Link>
        </li>
    );
};

export default NavBar;