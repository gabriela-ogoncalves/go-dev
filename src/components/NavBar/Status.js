import React, { useContext, useEffect } from 'react';

import Context from '../../Context';

import './styles.scss';

const Status = () => {
  const [isLogged] = useContext(Context);

  useEffect(() => {},[isLogged]); 

  return(
    <div className="user-status-menu">
      Ol√°, {isLogged.username}
    </div>
  );
};

export default Status;