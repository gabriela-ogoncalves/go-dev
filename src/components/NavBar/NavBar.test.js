import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import NavBar from './NavBar';

import '@testing-library/jest-dom';

describe('NavBar', () => {
  let navBar;

  beforeEach(() => {
    render(<NavBar />, { wrapper: BrowserRouter });
    navBar = screen.getByTestId('nav-bar');
  });

  it('Deve renderizar o componente normalmente', () => {
    expect(navBar).toMatchSnapshot();
  });

  it('Deve renderizar a logo do Go Dev', () => {
    const logo = navBar.querySelector('img');
    expect(logo).toBeInTheDocument();
    expect(logo).toHaveAttribute('class', 'logo-image');
    expect(logo).toHaveAttribute('alt', 'Logo Go Dev');
  });
});

describe('NavBar navegação', () => {
  let navBarNavigation, links;

  beforeEach(() => {
    render(<NavBar />, { wrapper: BrowserRouter });
    navBarNavigation = screen.getByTestId('nav-bar');
    links = screen.getAllByRole('link');
  });

  it('Deve encaminhar para a home', () => {
    const home = links[0];
    fireEvent.click(home);
    expect(home.toString()).toBe('http://localhost/');
  });

  it('Deve encaminhar para trilhas', () => {
    const trilhas = links[1];
    fireEvent.click(trilhas);
    expect(trilhas.toString()).toBe('http://localhost/trilhas');
  });

  it('Deve encaminhar para certificados', () => {
    const certificados = links[2];
    fireEvent.click(certificados);
    expect(certificados.toString()).toBe('http://localhost/certificados');
  });
});