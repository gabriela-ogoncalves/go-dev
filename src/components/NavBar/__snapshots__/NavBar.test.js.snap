// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar Deve renderizar o componente normalmente 1`] = `
<article
  data-testid="nav-bar"
>
  <nav
    class="navigation"
  >
    <a
      class="brand-name"
      data-testid="go-dev-logo"
      href="/"
    >
      <img
        alt="Logo Go Dev"
        class="logo-image"
        src="go-dev-logo.png"
      />
    </a>
    <button
      class="hamburger"
    >
      <svg
        class="h-5 w-5"
        fill="white"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3 5a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM9 15a1 1 0 011-1h6a1 1 0 110 2h-6a1 1 0 01-1-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <div
      class="navigation-menu"
    >
      <ul>
        <li
          class=""
        >
          <a
            href="/trilhas"
            role="link"
          >
            Trilhas
          </a>
        </li>
        <li
          class=""
        >
          <a
            href="/certificados"
            role="link"
          >
            Certificados
          </a>
        </li>
      </ul>
    </div>
  </nav>
</article>
`;
